platform-db-runtime/src/main/java/com/softicar/platform/db/runtime/table/validator/DbTableValidator.java

```diff
@@ -3,71 +3,145 @@
 import com.softicar.platform.common.core.utils.CastUtils;
 import com.softicar.platform.common.string.Imploder;
 import com.softicar.platform.db.runtime.field.IDbField;
+import com.softicar.platform.db.runtime.field.IDbForeignRowField;
 import com.softicar.platform.db.runtime.field.IDbStringField;
+import com.softicar.platform.db.runtime.key.IDbKey;
 import com.softicar.platform.db.runtime.table.IDbTable;
 import com.softicar.platform.db.sql.type.SqlFieldType;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.HashSet;
 import java.util.List;
+import java.util.stream.Collectors;
 import org.junit.Assert;
 
 public class DbTableValidator<R> extends Assert {
 
 	private final IDbTable<R, ?> table;
 	private final List<? extends IDbField<R, ?>> dataFields;
+	private final Collection<? extends IDbKey<R>> allKeys;
 
 	public DbTableValidator(IDbTable<R, ?> table) {
 
 		this.table = table;
-		this.dataFields = this.table.getDataFields();
+		this.dataFields = table.getDataFields();
+		this.allKeys = table.getAllKeys();
 	}
 
 	public void validate() {
 
-		validateStringFields();
-		validateNullability();
+		var errors = new ErrorList();
+		errors.addAll(validateStringFields());
+		errors.addAll(validateNullability());
+		errors.addAll(validatePrimaryKey());
+		errors.addAll(validateForeignKeys());
+
+		if (!errors.isEmpty()) {
+			throw new AssertionError(Imploder.implode(errors, "\n"));
+		}
 	}
 
-	private void validateStringFields() {
+	private List<String> validateStringFields() {
 
-		Collection<String> errors = new ArrayList<>();
+		var errors = new ErrorList();
 
 		for (IDbField<R, ?> field: dataFields) {
 			if (field.getFieldType().equals(SqlFieldType.STRING)) {
 				IDbStringField<R> stringField = CastUtils.cast(field);
 				if (stringField.getMaximumLength() == 0 && stringField.getLengthBits() == 0) {
-					String message = String//
-						.format(//
+					errors
+						.add(//
 							"String field %s.`%s` needs either a 'MAXLENGTH' or a 'LENGTHBITS' attribute.",
 							table.getFullName(),
-							field.getTitle());
-					errors.add(message);
+							field.getName());
 				}
 			}
 		}
 
-		if (!errors.isEmpty()) {
-			throw new AssertionError(Imploder.implode(errors, "\n"));
-		}
+		return errors;
 	}
 
-	private void validateNullability() {
+	private List<String> validateNullability() {
 
-		Collection<String> errors = new ArrayList<>();
+		var errors = new ErrorList();
 
 		for (IDbField<R, ?> field: dataFields) {
 			if (field.hasDefault() && field.getDefault() == null && !field.isNullable()) {
-				String message = String//
-					.format(//
+				errors
+					.add(//
 						"Field %s.`%s` defaults to NULL but it is not NULLABLE.",
 						table.getFullName(),
-						field.getTitle());
-				errors.add(message);
+						field.getName());
+			}
+
+			if (field.isNullable() && !field.hasDefault()) {
```

Comment:
what is the rationale that nullable fields need a default?

Reply:
How would I know? Quoting the MySQL docs at https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html#data-type-defaults-implicit:
>  If the column can take NULL as a value, the column is defined with an explicit DEFAULT NULL clause.

Reply:
pfff, ok ;)